android/java/org/chromium/chrome/browser/crypto_wallet/activities/AddAssetActivity.java

```diff
@@ -0,0 +1,258 @@
+    private static final String TAG = "AddAssetActivity";
+
+    private static final String NFT_CUSTOM_ASSET = "nftCustomAsset";
+
+    private WalletModel mWalletModel;
+    private NetworkModel mNetworkModel;
+    private NetworkInfo mNetworkInfo;
+
+    private boolean mNftsOnly;
+    private boolean mPauseTextWatcher;
+
+    private Toolbar mToolbar;
+    private Spinner mNetworkSpinner;
+    private View mTokenIdSection;
+    private View mTokenDecimalSection;
+    private TextView mTvDecimalTitle;
+    private EditText mTokenDecimalsEdit;
+    private TextView mTvAddressTitle;
+    private NetworkSpinnerAdapter mNetworkAdapter;
+    private Button mAdd;
+    private EditText mTokenNameEdit;
+    private EditText mTokenContractAddressEdit;
+    private EditText mTokenSymbolEdit;
+    private EditText mTokenIdEdit;
+
+    /**
+     * Gets intent to start new activity {@code AddAssetActivity}.
+     * @param context Android context required to instantiate the intent.
+     * @param nftOnly {@code true} if the activity will add a new NFT,
+     * {@code false} to add a new custom asset.
+     * @return Intent to start new activity {@code AddAssetActivity}.
+     */
+    public static Intent getIntent(@NonNull final Context context, final boolean nftOnly) {
+        Intent intent = new Intent(context, AddAssetActivity.class);
+        intent.putExtra(NFT_CUSTOM_ASSET, nftOnly);
+        return intent;
+    }
+
+    @Override
+    protected void onPreCreate() {
+        Intent intent = getIntent();
+        if (intent != null) {
+            mNftsOnly = intent.getBooleanExtra(NFT_CUSTOM_ASSET, false);
+        }
+    }
+
+    @Override
+    protected void triggerLayoutInflation() {
+        setContentView(R.layout.activity_add_asset);
+
+        mToolbar = findViewById(R.id.toolbar);
+        mToolbar.setTitle(mNftsOnly ? R.string.wallet_add_nft : R.string.wallet_add_custom_asset);
+        setSupportActionBar(mToolbar);
+        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
+
+        mNetworkSpinner = findViewById(R.id.network_spinner);
+        mNetworkAdapter = new NetworkSpinnerAdapter(this, Collections.emptyList());
+        mNetworkSpinner.setAdapter(mNetworkAdapter);
+        mTokenIdSection = findViewById(R.id.token_id_section);
+        mTokenDecimalSection = findViewById(R.id.token_decimal_section);
+        mTvDecimalTitle = findViewById(R.id.token_decimals_title);
+        mTokenDecimalsEdit = findViewById(R.id.token_decimals);
+        mTvAddressTitle = findViewById(R.id.token_contract_address_title);
+
+        mAdd = findViewById(R.id.add);
+        mTokenNameEdit = findViewById(R.id.token_name);
+        mTokenContractAddressEdit = findViewById(R.id.token_contract_address);
+        mTokenSymbolEdit = findViewById(R.id.token_symbol);
+        mTokenIdEdit = findViewById(R.id.token_id);
+
+        onInitialLayoutInflationComplete();
+    }
+
+    @Override
+    public void onStartWithNative() {
```

Comment:
why did you pick `onStartWithNative`? I'm not sure does it get called after the native inited. I see an assert in a base class. We always used `public void finishNativeInitialization() {` to start native work and it works ok. Maybe let's switch to it? It could be `onStartWithNative` is also ok but I'm not certain.

Reply:
My understanding is that `onStartWithNative` is the only method that can guarantee `onStart` + native ready.
The main problem with `finishNativeInitialization` is that it's triggered independently of layout inflation: thus it may end up in a case where native is ready and layout inflation is not. Especially for those devices that don't support instant start.
I've had a look at the chromium source code and some of the logic is explained in here in `AsyncInitialization`:
https://source.chromium.org/chromium/chromium/src/+/main:chrome/android/java/src/org/chromium/chrome/browser/init/AsyncInitializationActivity.java;l=174-192;drc=144ccdb824e9edf088eb5c4f5b6c43ac5803df7b

For the specific case where we need layout inflation completed + native ready the method `onStartWithNative` seems to be the perfect match: https://source.chromium.org/chromium/chromium/src/+/main:chrome/android/java/src/org/chromium/chrome/browser/init/ChromeActivityNativeDelegate.java;l=25-27;drc=144ccdb824e9edf088eb5c4f5b6c43ac5803df7b?q=chromeActivityNativeDelegate&ss=chromium%2Fchromium%2Fsrc

Per documentation, it says `Carry out native code dependent` (so native ready guaranteed), `during on Activity.onStart()` (so layout ready as `onStart` can be called only after  `onCreate` where the layout is inflated).

Let me know if my investigation was inaccurate, or maybe I'm missing some piece of the puzzle. Happy to fall back to `finishNativeInitialization` method if `onStartWithNative` is not the best method to use here.

Reply:
One downside of `onStartWithNative` is that it can be called again based on activity state so can use a boolean to avoid re-initialisation though better if we can find a one-shot method/way like `onCreate` etc.(as onCreateNative is async), with native safe calls (if required).

Reply:
Yes, good point. I would introduce a boolean as a safety net.
Unfortunately, I haven't found any other callback that could guarantee [native ready + post layout inflation].
`onCreateWithNative` @Pavneet-Sing would have been a great candidate for that, but the Chromium team made the overriding call as `final` : https://source.chromium.org/chromium/chromium/src/+/main:chrome/android/java/src/org/chromium/chrome/browser/init/AsyncInitializationActivity.java;l=568;bpv=1;bpt=1?q=AsyncInitialization&ss=chromium%2Fchromium%2Fsrc

So, in our activities extending `AsyncInitializationActivity` we cannot use it.

Reply:
I've added a boolean as an extra precaution, but it's very unlikely to happen as the Wallet gets closed every time the activity goes in background. We should be super safe.